[Toh Wei Jie] iP

TohWeiJie01 opened the pull request:
# DUKE
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes/))

Duke frees your mind of having to remember things you need to do. It's,
* text-based
* easy to learn
* ~~FAST~~ SUPER FAST to use

All you need to do is,
1. Download it from [here](https://github.com/TohWeiJie01/ip/releases/tag/A-Jar)
2. Double-click it
3. Add your tasks
4. Let it manage your tasks for you! :baguette_bread::rofl::baguette_bread:

Did I mention that it is __FREE__!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)





 If you are a Java programmer, you can use it to practice Java too! Here's the `main` method:

``` java
public static void main(String[] args) throws IOException {

    Duke duke = new Duke(PATH_NAME);
    duke.run();

}
```

Commit:
Add Gradle support

TohWeiJie01 committed:
Improved Logo for Duke.java and added private fields and constructor to initialise Duke in Duke.java

TohWeiJie01 committed:
added method: commanding() to Duke.java and also initialised Duke in the main method and called upon the commanding method

TohWeiJie01 committed:
added javadocs for the commanding method

TohWeiJie01 committed:
changed the private field of Duke to be an array of size 100 to act as the to-do list

TohWeiJie01 committed:
added the store method to Duke.java and also added a pointer nextSpaceToStore, along with more javadocs updating

TohWeiJie01 committed:
added the list method to Duke.java, edited the commanding method to check for the command "list" and updated javadocs

TohWeiJie01 committed:
added Task.java to encapsulate a task

TohWeiJie01 committed:
Added TaskList.java to encapsulate a list of tasks and edited Duke.java to fit the logic of a TaskList (for OOP)

TohWeiJie01 committed:
added the feature of done task

TohWeiJie01 committed:
Added Todo.java, Deadline.java and Event.java

TohWeiJie01 committed:
Edited in Duke.java to look for the "todo", "deadline" and "event" commands and initialise the specific type of tasks

TohWeiJie01 committed:
added Automated Text UI Testing and adjusted code to meet expected

TohWeiJie01 committed:
Deleted unnecessary code

TohWeiJie01 committed:
Added DukeException.java to handle Duke specific exceptions

TohWeiJie01 committed:
Handled unrecognised command

TohWeiJie01 committed:
Handled empty todo, deadline and event commands and updated testing files

TohWeiJie01 committed:
Utillised Java Collections - ArrayList instead of a normal array

TohWeiJie01 committed:
Added the method to delete tasks from the TaskList

TohWeiJie01 committed:
Edited the input and expected testing

TohWeiJie01 committed:
Edit the names of variables, methods and Javadoc

TohWeiJie01 committed:
Add Storage.java

TohWeiJie01 committed:
Update Duke.java to have the load and save function

TohWeiJie01 committed:
Refactor code

TohWeiJie01 committed:
Update Deadline.java and Event.java with DateTime feature

TohWeiJie01 committed:
Update Duke.java with the Date and Time feature

TohWeiJie01 committed:
Merge branch 'branch-Level-7'

TohWeiJie01 committed:
Merge branch 'branch-Level-8'

TohWeiJie01 committed:
Update code to have Date and Time functionality

TohWeiJie01 committed:
Add more OOP

TohWeiJie01 committed:
Divide classes into packages

TohWeiJie01 committed:
Create DukeTest.java

TohWeiJie01 committed:
Add JUnit tests

TohWeiJie01 committed:
Add and edit JavaDoc comments to code

TohWeiJie01 committed:
Follow the Coding Standard

TohWeiJie01 committed:
Add FindCommand.java

TohWeiJie01 committed:
Update code to have Find functionality

TohWeiJie01 committed:
Merge branch 'branch-A-JavaDoc'

TohWeiJie01 committed:
Merge branch 'branch-A-CodingStandard'

TohWeiJie01 committed:
Merge branch 'branch-Level-9'

TohWeiJie01 committed:
Update Coding Standard in Parser.java

xiongjya commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,39 @@
+ */
+public class DeleteCommand extends Command {
+    /** The delete command inputted by the user. */
+    String deleteDescription;
```
Do you think that there is a better name alternative for this variable? The current name sounds more like an action instead of a noun.

Morrow1ndy replied:
Perhaps the current naming is not appropriate for a string variable?

xiongjya commented on src/main/java/duke/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,55 @@
+ */
+public class DeadlineCommand extends Command {
+    /** The deadline command inputted by the user. */
+    String fullCommand;
```
Perhaps it is better to specify the access modifier for fullCommand in this case? I noticed the same issue in several other places too, such as in the DeleteCommand, DoneCommand, EventCommand and a few more classes.

xiongjya commented on src/main/java/duke/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,55 @@
+            String deadlineDescription = fullCommand.split(" ", 2)[1];
+            int pos = deadlineDescription.indexOf("/");
+            String description = deadlineDescription.substring(0, pos - 1);
+            String by = deadlineDescription.substring(pos + 4);
```
Would it be better to name this as 'duedate' instead? If I am not wrong, the variable represents a date/noun, while 'by' might not be the most suitable name in this case. Similarly, I have noticed that you have used this name at some other areas as well.

xiongjya commented on src/main/java/duke/command/EventCommand.java:
```diff
@@ -0,0 +1,55 @@
+            String eventDescription = fullCommand.split(" ", 2)[1];
+            int pos = eventDescription.indexOf("/");
+            String description = eventDescription.substring(0, pos - 1);
+            String at = eventDescription.substring(pos + 4);
```
Similarly, instead of 'at', which is a preposition, would it be better to name the variable (that is a noun), 'eventTime'?

xiongjya commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,29 @@
+ */
+public abstract class Command {
+    /**
+     * The abstract method to execute the correct response to the user's command.
```
To better follow the coding standards, maybe you could add an empty line between the description and parameter section?

xiongjya reviewed (commented):
LGTM 👍 
Overall, some cosmetic changes requested. I like how you created classes for each type of command as well, but do ensure that the total number of classes/files is manageable!

TohWeiJie01 committed:
Edit code to coding standard and other coding quality guidelines

TohWeiJie01 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

TohWeiJie01 committed:
Enable Gradle to run

TohWeiJie01 committed:
Merge branch 'branch-A-Gradle'

TohWeiJie01 committed:
Add Checkstyle to Gradle

TohWeiJie01 committed:
Merge branch 'branch-A-Checkstyle'

SajalVaishnav commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,32 @@
+package duke.command;
```
good use of packages!

SajalVaishnav commented on src/main/java/duke/command/EventCommand.java:
```diff
@@ -0,0 +1,58 @@
+     * @param ui The Ui Duke utilises to interact with the user.
+     */
+    @Override
+    public void execute(Storage storage, TaskList taskList, Ui ui) throws IOException, DukeException {
```
The execute in many of the commands does many similar things, and maybe has the responsibility to do too many things; perhaps making the flow more modular could be nice?

SajalVaishnav commented on src/main/java/duke/exception/DukeException.java:
```diff
@@ -0,0 +1,26 @@
+     *
+     * @param message The message to be shown should there be a Duke Exception.
+     */
+    public DukeException(String message) {
```
Maybe a couple extra exception types can help handle exceptional situations with more tact?

SajalVaishnav commented on src/test/java/duke/task/DeadlineTest.java:
```diff
@@ -0,0 +1,14 @@
+
+public class DeadlineTest {
+    @Test
+    public void testDeadline() {
```
maybe adding the input type to the function name will be more adhering to the module code standards?

TohWeiJie01 committed:
Add GUI for Duke

TohWeiJie01 committed:
Use FXML

TohWeiJie01 committed:
Merge branch 'branch-Level-10'

Morrow1ndy commented on src/main/java/duke/command/DoneCommand.java:
```diff
@@ -0,0 +1,39 @@
+ */
+public class DoneCommand extends Command {
+    /** The done command inputted by the user. */
+    String doneDescription;
```
Maybe the same naming issue here?

Morrow1ndy commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,127 @@
+    private File file;
+
+    /** Constant String for the symbol to separate the descriptions of the saved tasklist. */
+    private final String SEPARATE = " / ";
```
Maybe the naming of the string here should be a noun?

Morrow1ndy commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,127 @@
+            while (scanner.hasNextLine()) {
+                String currTask = scanner.nextLine();
+                Task task = null;
+                String[] taskBreakdown = currTask.split(SEPARATE);
```
Maybe the naming of the array here should be something to suggest the data structure used is an array?

Morrow1ndy reviewed (commented):
Overall your code is very easy to read and follow!

TohWeiJie01 committed:
Add new exception classes that inherit from DukeException

There is only a singular DukeException class to account for all the possible exceptions specific to Duke.

However, this is very restrictive and not a good style of coding by OOP rules.

In order to improve this exception issue, three new separate exception classes that inherit from DukeException is created. One handles the case where the given input is completely unrecognised, another handles the case where the description of the task we are adding is missing and the last handles the case for inputs that require an index but do not have it.

By separating the possible exceptions into separate classes, it is now clearer exactly which exception occurred and what is exactly the problem should it arise in Duke.

TohWeiJie01 committed:
Add enums for the different task types

There is no usage of enums, hence the todo, deadline and event tasks all
had to be separately dealt with, despite having similar behaviors.

The common behaviors across the Task types could cause code
duplications.

Creating an enum class to represent all three task types allows us to
check for the different task types and utilise the same method to handle
all of them, while still being able to differentiate them.

Using enums is preferable as there is only a fixed set of task types.

TohWeiJie01 committed:
Add more enum and exception classes

TohWeiJie01 committed:
Update code to include new exceptions and enums

TohWeiJie01 committed:
Edit GUI

TohWeiJie01 committed:
Add assertions

TohWeiJie01 committed:
Update to improve Code Quality

TohWeiJie01 committed:
Update assertions

TohWeiJie01 committed:
Merge pull request #1 from TohWeiJie01/branch-A-Assertions

Branch-A-Assertions

TohWeiJie01 committed:
Merge branch 'master' of https://github.com/TohWeiJie01/ip

TohWeiJie01 committed:
Merge branch 'master' into branch-A-CodeQuality

TohWeiJie01 committed:
Merge pull request #2 from TohWeiJie01/branch-A-CodeQuality

branch-A-CodeQuality

TohWeiJie01 committed:
Edit tasklist

TohWeiJie01 committed:
Merge branch 'master' of https://github.com/TohWeiJie01/ip

TohWeiJie01 committed:
Add detect duplicates functionality

TohWeiJie01 committed:
Merge branch 'branch-A-CodeQuality'

TohWeiJie01 committed:
Edit GUI

TohWeiJie01 committed:
Add Ui image

TohWeiJie01 committed:
Edit Gui and Ui image

TohWeiJie01 committed:
Add User Guide

TohWeiJie01 committed:
Edit Javadocs

TohWeiJie01 committed:
Add welcome message upon Gui startup and automated closure of Gui

TohWeiJie01 committed:
Edit Gui

TohWeiJie01 committed:
Edit Gui and to make it compatible with other platforms

TohWeiJie01 committed:
Edit Javadocs

TohWeiJie01 committed:
Add assertions

TohWeiJie01 committed:
Update user guide

TohWeiJie01 committed:
Edit code for DeadlineCommand and EventCommand

TohWeiJie01 committed:
Remove unnecessary Junit Test
